docs/proposals/20230117-cluster-level-identity.md

```diff
@@ -0,0 +1,162 @@
+}
+```
+The following custom resources will be added.
+
+```go
+
+type PrincipalType string
+
+const (
+    // UserPrincipal represents a user princpal.
+    UserPrincipal PrincipalType = "UserPrincipal"
+)
+
+// OCIClusterIdentitySpec defines the parameters that are used to create an OCIIdentity.
+type OCIClusterIdentitySpec struct {
+  // Type is the type of OCI Principal used.
+  // UserPrincipal is the only supported value
+  Type PrincipalType `json:"type"`
+  
+  // PrincipalSecret is a secret reference which contains the authentication credentials for the principal.
+  // +optional
+  PrincipalSecret corev1.SecretReference `json:"clientSecret,omitempty"`
+
+  // AllowedNamespaces is used to identify the namespaces the clusters are allowed to use the identity from.
+  // Namespaces can be selected either using an array of namespaces or with label selector.
+  // An empty allowedNamespaces object indicates that OCIClusters can use this identity from any namespace.
+  // If this object is nil, no namespaces will be allowed (default behaviour, if this field is not provided)
+  // A namespace should be either in the NamespaceList or match with Selector to use the identity.
+  //
+  // +optional
+  // +nullable
+  AllowedNamespaces *AllowedNamespaces `json:"allowedNamespaces"`
+}
+
+// OCIClusterIdentityStatus defines the observed state of OCIClusterIdentity.
+type OCIClusterIdentityStatus struct {
+  // Conditions defines current service state of the OCIClusterIdentity.
+  // +optional
+  Conditions clusterv1.Conditions `json:"conditions,omitempty"`
+}
+
+type OCIClusterIdentity struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  Spec   OCIClusterIdentitySpec   `json:"spec,omitempty"`
+  Status OCIClusterIdentityStatus `json:"status,omitempty"`
+}
+
+```
+
+
+### Example Cluster template
+
+```yaml
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: OCICluster
+spec:
+  identityRef:
+    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+    kind: OCIClusterIdentity
+    name: <test-identity>
+    namespace: <test>
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: OCIClusterIdentity
+metadata:
+  name: test-identity
+  namespace: test
+spec:
+  type: UserPrincipal
+  clientSecret: 
+    name: user-credentials
+    namespace: secret-namespace
+  allowedNamespaces:
+    list:
+      - test
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: user-credentials
+  namespace: secret-namespace
+type: Opaque
+data:
+  tenancy: <>
+  user: <>
+  passphrase: <> -> optional
+  key: <>
+  fingerprint: <>
+---
+```
+
+### Implementation Details
+
+#### Controller changes
+
+Cluster and machine controller will lookup if the cluster has an associated identity. If an identity exists
+OCI clients will be created using the corresponding identity principals. If identity does not exist
+the old behaviour of using the CAPOCI pod level OCI identity principals will still be used.
```

Comment:
So the default identity principals will still be required for initialization of CAPOCI correct?

Reply:
currently yes, we may want to make it optional as well, but that will be a separate story.